docs/architecture.md

```diff
@@ -71,9 +71,27 @@ Pay attention especially to the following sections:
 - **Architecture Invariant**: important note about crucial assumptions about the code/behavior
 - **API Boundary**: an exposed interface that requires careful design, e.g. because it has internal users to support
 
+### `ast`
+
+Types that serve as the Gleam AST, both in its *untyped* and *typed* form.
+Each tree describes a single Gleam module at its root.
+
+Untyped trees contain untyped expressions only.
+This form purely describes a sequence of characters.
+
+Typed trees contain typed expressions together with module metadata.
+Here, code has gained the context necessary to produce final Gleam output.
+
 ### `parse`
 
-TODO
+The parser receives source code for a single Gleam module (as `&str`, as opposed to a stream)
+and produces an untyped AST together with `ModuleExtra`
+containing various types of comments plus empty lines.
+
+Currently this means that parsing is tightly coupled with the AST structure.
+This might prove inflexible when creating tools such as a Gleam LSP.
+
```

Comment:
```suggestion
```

I think speculation about future features belongs in the issues and discussions for those features.

Reply:
Agreed, my intention was to give an example where this coupling falls short when working with code.

Reply:
Replaced the LSP remark with an API boundary note about high-level operations